File: 프로그래머스/2/42885. 구명보트/구명보트.cpp
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<int> people, int limit) {
    sort(people.begin(), people.end());
    
    int cnt = 0; // Initialize boat count
    int left = 0, right = people.size() - 1; // Initialize pointers for heaviest and lightest people
    
    while (left <= right) {
        if (people[left] + people[right] <= limit) {
            left++; // Move the lighter person to the next boat
        }
        right--; // Move the heavier person to the next boat in any case
        cnt++; // Increment boat count
    }
    
    return cnt;
}

File: 프로그래머스/2/42885. 구명보트/구명보트.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int[] people, int limit) {
        // 사람 몸무게 people
        // 구명보트 무게제한 limit
        
        // 정렬
        Arrays.sort(people);
        
        // 투포인터
        int lo = 0;
        int hi = people.length - 1;
        int answer = 0;
        
        while(lo <= hi){
            
            // 겹치는 경우
            if(lo == hi){
                answer++;
                break;
            }
            
            // 합계
            int sum = people[lo] + people[hi];
            
            // 합계가 초과하면 한명만 탈출
            if(sum > limit){
                hi--;
            }
            // 2명 가능
            else{
                lo++;
                hi--;
            }
            // 개수 증가
            answer++;
            
        }
        
        return answer;
    }
}